Check overflows and validate timestamp9 value range.

higuoxing opened the pull request:
This patch helps add checks for potential overflows when converting other date time types to timestamp9 and adds checks for validating then timestamp9 values range [1900-01-01 00:00:00.000000000 UTC+0, 2262-01-01 00:00:00.000000000 UTC+0).

Trying to fix #12 

P.S. I don't know if the value range is appropriate, let's discuss about it?

higuoxing committed:
Check overflows and validate timestamp9 value range.

This patch helps add checks for potential overflows when converting
other date time types to timestamp9 and adds checks for validating then
timestamp9 values range [1900-01-01 00:00:00.000000000 UTC+0, 2262-01-01
00:00:00.000000000 UTC+0).

fvannee commented:
Thanks for the submission. I'm open to having better checks here to prevent cases like the ones mentioned in #12 . Did you pick the current min/max range arbitrarily? I think the range is well suited for most cases, but maybe it's too restrictive. Would it be difficult to support the full range of `bigint` from minimum value to maximum value or do you think it'd be too complicated with overflow checks?

fvannee commented:
`timestamp9_recv` would need a check added too, otherwise invalid values can be constructed from the binary libpq protocol

higuoxing committed:
Trying to address review comments.

- Validate timestamp9 value in timestamp9_recv().
- Enlarge the timestamp9 value range to '[1700-01-01 UTC+0, 2262-01-01
  UTC+0)'.

higuoxing commented:
> Did you pick the current min/max range arbitrarily? I think the range is well suited for most cases, but maybe it's too restrictive. Would it be difficult to support the full range of `bigint` from minimum value to maximum value or do you think it'd be too complicated with overflow checks?

Yes, I pick the min/max year arbitrarily, but pick the date `01-01` deliberately, since it's easy for validating the timestamp and that's what PostgreSQL do for the builtin timestamp range.

Based on your comment, I enlarge the value range to `[1700-01-01 00:00:00.000000000 UTC+0, 2262-01-01 00:00:00.000000000 UTC+0)` in my latest patch, because `INT64_MIN` corresponds to `1677-09-21 00:12:43.145224192 +0000` and `INT64_MAX` corresponds to `2262-04-11 23:47:16.854775807 +0000`.

> `timestamp9_recv` would need a check added too, otherwise invalid values can be constructed from the binary libpq protocol

Good catch! Thanks for pointing it out.

higuoxing committed:
Validate timestamp range when doing bigint->timestamp9 conversion.

asiayeah commented:
I don't see it merged into 1.4.0.

Any plan to merge and release this PR? Thank you.
